cpp/src/arrow/compute/kernels/take.cc

```diff
@@ -0,0 +1,221 @@
+Status TakeImpl(FunctionContext* context, const ValueArray& values,
+                const IndexArray& indices, OutBuilder* builder) {
+  if (values.null_count() == 0) {
+    return TakeImpl<OutOfBounds, true>(context, values, indices, builder);
+  }
+  return TakeImpl<OutOfBounds, false>(context, values, indices, builder);
+}
+
+template <typename ValueArray, typename IndexArray, typename OutBuilder>
+Status TakeImpl(FunctionContext* context, const ValueArray& values,
+                const IndexArray& indices, const TakeOptions& options,
+                OutBuilder* builder) {
+  switch (options.out_of_bounds) {
+    case TakeOptions::ERROR:
+      return TakeImpl<TakeOptions::ERROR>(context, values, indices, builder);
+    case TakeOptions::TONULL:
+      return TakeImpl<TakeOptions::TONULL>(context, values, indices, builder);
+    case TakeOptions::UNSAFE:
+      return TakeImpl<TakeOptions::UNSAFE>(context, values, indices, builder);
+    default:
+      ARROW_LOG(FATAL) << "how did we get here";
+      return Status::NotImplemented("how did we get here");
+  }
+}
+
+template <typename IndexType>
+struct UnpackValues {
+  using IndexArrayRef = const typename TypeTraits<IndexType>::ArrayType&;
+
+  template <typename ValueType>
+  Status Visit(const ValueType&) {
+    using ValueArrayRef = const typename TypeTraits<ValueType>::ArrayType&;
+    using OutBuilder = typename TypeTraits<ValueType>::BuilderType;
+    IndexArrayRef indices = static_cast<IndexArrayRef>(*params_.indices);
+    ValueArrayRef values = static_cast<ValueArrayRef>(*params_.values);
+    std::unique_ptr<ArrayBuilder> builder;
+    RETURN_NOT_OK(MakeBuilder(params_.context->memory_pool(), values.type(), &builder));
+    RETURN_NOT_OK(builder->Reserve(indices.length()));
+    RETURN_NOT_OK(TakeImpl(params_.context, values, indices, params_.options,
+                           static_cast<OutBuilder*>(builder.get())));
+    return builder->Finish(params_.out);
+  }
+
+  Status Visit(const NullType& t) {
+    auto indices_length = params_.indices->length();
+    if (params_.options.out_of_bounds == TakeOptions::ERROR && indices_length != 0) {
+      auto indices = static_cast<IndexArrayRef>(*params_.indices).raw_values();
+      auto max = *std::max_element(indices, indices + indices_length);
+      if (static_cast<int64_t>(max) > params_.values->length()) {
+        return Status::Invalid("out of bounds index");
+      }
+    }
+    params_.out->reset(new NullArray(indices_length));
+    return Status::OK();
+  }
+
+  Status Visit(const DictionaryType& t) {
+    auto dictionary_indices = params_.values->data()->Copy();
+    dictionary_indices->type = t.index_type();
+    TakeParameters params = params_;
+    params.values = MakeArray(dictionary_indices);
+    UnpackValues<IndexType> unpack = {params};
+    RETURN_NOT_OK(VisitTypeInline(*t.index_type(), &unpack));
+    (*params_.out)->data()->type = dictionary(t.index_type(), t.dictionary());
```

Comment:
This is likely wrong, the user could assume to upcast this into DictionaryArray, yet it's a NumericArray (since it'll be build with a NumericBuilder).

Note this currently works because Datum(array) just takes `array->data`. For the sake of completeness or future-proof, either make param.out an ArrayData, or properly return a DictionnaryArray.